File: app/src/main/java/com/app/cheffyuser/profile/fragments/AccountFragment.java
package com.app.cheffyuser.profile.fragments;


import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.location.Address;
import android.location.Geocoder;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.fragment.app.Fragment;
import androidx.viewpager.widget.ViewPager;

import com.app.cheffyuser.R;
import com.app.cheffyuser.networking.Constant;
import com.app.cheffyuser.profile.TabsAdapter;
import com.app.cheffyuser.profile.activities.EditProfileActivity;
import com.app.cheffyuser.profile.utils.Utils;
import com.bumptech.glide.Glide;
import com.google.android.material.tabs.TabLayout;

import java.io.IOException;
import java.util.List;
import java.util.Locale;

/**
 * A simple {@link Fragment} subclass.
 */
public class AccountFragment extends Fragment {



    public AccountFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(final LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view= inflater.inflate(R.layout.fragment_account, container, false);

        ImageView imgProfileEdit=view.findViewById(R.id.img_profile_edit);
        ImageView imgUser=view.findViewById(R.id.user_image);


        TabLayout tabLayout = (TabLayout)view. findViewById(R.id.profile_tab_layout);
        tabLayout.addTab(tabLayout.newTab().setText("Your Favorite"));
        tabLayout.addTab(tabLayout.newTab().setText("Account Setting"));
        tabLayout.setTabGravity(TabLayout.GRAVITY_FILL);


        TextView txtUserName=view.findViewById(R.id.tv_user_name);
        TextView txtAddress=view.findViewById(R.id.tv_user_address);

        //Fetching cell from shared preferences
        SharedPreferences sp;
        sp =getContext().getSharedPreferences(Constant.SHARED_PREF_NAME, Context.MODE_PRIVATE);
        String name = sp.getString(Constant.SP_NAME, "N/A");
        String image_path=sp.getString(Constant.SP_IMAGE_PATH,"N/A");

        String latitude=sp.getString(Constant.SP_LOCATION_LAT,"");
        String longitude=sp.getString(Constant.SP_LOCATION_LON,"");


        txtUserName.setText(name);

        final String getAddress= Utils.getCurrentPlace(getActivity(),latitude,longitude);
        txtAddress.setText(getAddress);

        //Loading image from url into imageView
        Glide.with(getActivity())
                .load(image_path)
                .placeholder(R.drawable.image_placeholder)
                .error(R.drawable.image_placeholder)
                .into(imgUser);





        final ViewPager viewPager =(ViewPager)view.findViewById(R.id.profile_view_pager);
        TabsAdapter tabsAdapter = new TabsAdapter(getActivity().getSupportFragmentManager(), tabLayout.getTabCount());
        viewPager.setAdapter(tabsAdapter);
        viewPager.addOnPageChangeListener(new TabLayout.TabLayoutOnPageChangeListener(tabLayout));
        tabLayout.setOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                viewPager.setCurrentItem(tab.getPosition());
            }
            @Override
            public void onTabUnselected(TabLayout.Tab tab) {

            }
            @Override
            public void onTabReselected(TabLayout.Tab tab) {

            }
        });



        imgProfileEdit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent=new Intent(getActivity(), EditProfileActivity.class);
                intent.putExtra("address",getAddress);
                startActivity(intent);

            }
        });
        return view;
    }


//
//    private String getCurrentPlace(String lat,String lon)
//    {
//
//        double latitude=Double.parseDouble(lat);
//        double longitude=Double.parseDouble(lon);
//        Geocoder geocoder;
//        List<Address> addresses = null;
//        geocoder = new Geocoder(getContext(), Locale.getDefault());
//
//        try {
//            addresses = geocoder.getFromLocation(latitude, longitude, 1); // Here 1 represent max location result to returned, by documents it recommended 1 to 5
//        } catch (IOException e) {
//            e.printStackTrace();
//        }
//
//        //for getting current location using gps
//        //https://developer.android.com/reference/android/location/Address
//        String address = addresses.get(0).getAddressLine(0); // If any additional address line present than only, check with max available address lines by getMaxAddressLineIndex()
//
////        String city = addresses.get(0).getLocality();
////        String state = addresses.get(0).getAdminArea();
////        String country = addresses.get(0).getCountryName();
//        // String postalCode = addresses.get(0).getPostalCode();
//
//        // String knownName = addresses.get(0).getFeatureName(); // Only if available else return NULL
//
//        //txtLocation.setText(address);
//        //Toast.makeText(this, "Post Code:"+postalCode+" "+knownName, Toast.LENGTH_SHORT).show();
//
//        return address;
//    }

}

File: app/src/main/java/com/app/cheffyuser/profile/fragments/AccountFragment.kt
package com.app.cheffyuser.profile.fragments


import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.app.cheffyuser.BuildConfig
import com.app.cheffyuser.R
import com.app.cheffyuser.create_account.model.ShippingRequest
import com.app.cheffyuser.home.fragments.BaseFragment
import com.app.cheffyuser.home.viewmodel.HomeViewModel
import com.app.cheffyuser.networking.Status
import com.app.cheffyuser.profile.activities.EditProfileActivity
import com.app.cheffyuser.profile.activities.ShippingActivity
import com.app.cheffyuser.profile.adapter.ProfilebarAdapter
import com.app.cheffyuser.utils.createSnack
import com.app.cheffyuser.utils.loadUrl
import com.google.android.material.tabs.TabLayout
import kotlinx.android.synthetic.main.fragment_account.*
import kotlinx.android.synthetic.main.fragment_user_main.*


/**
 * A simple [Fragment] subclass.
 */
class AccountFragment : BaseFragment() {

    private val vm: HomeViewModel by lazy {
        ViewModelProviders.of(activity!!).get(HomeViewModel::class.java)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_account, container, false)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        uiStuff()
        initTablayout()
    }

    private fun uiStuff() {

        vm.fetchUser().observe(this, Observer { dt ->
            val datas = dt.data

            when (dt.status) {
                Status.ERROR -> {

                    if (BuildConfig.DEBUG)
                        createSnack(ctx = activity!!, txt = "User can't be fetched now")

                }
                Status.SUCCESS -> {
                    tokenManager.userData = datas?.data

                    datas.let {

                        tv_user_name.text = it?.data?.name
                        user_image.loadUrl(it?.data?.imagePath, R.drawable.avatar_placeholder)
                    }
                }
                Status.LOADING -> {
                    //still loading data
                }
            }

        })

        vm.fetchShipping().observe(this, Observer {
            val datas = it.data?.shippingResponseData

            when (it.status) {
                Status.ERROR -> {

                    if (BuildConfig.DEBUG)
                        createSnack(
                            ctx = activity!!,
                            txt = "Debug only: Address can't be fetched now"
                        )
                }
                Status.SUCCESS -> {
                    if (!datas.isNullOrEmpty()) {

                        val data = datas.first { d -> d.isDefaultAddress == true }

                        tokenManager.shippingData = data
                        vm.shippingData.value = data

                        val ship = ShippingRequest()
                        ship.lat = data.lat
                        ship.lon = data.lon
                        ship.state = data.state
                        ship.city = data.city
                        ship.addressLine1 = data.addressLine1
                        ship.addressLine2 = data.addressLine2
                        ship.zipCode = data.zipCode

                        tokenManager.shippingData2 = ship

                    }

                    checkShippingData()

                    datas.let {
                        if (!datas.isNullOrEmpty())
                            tv_user_address.text = "${it?.get(0)?.city} ${it?.get(0)?.state}"
                    }
                }
                Status.LOADING -> {
                    //still loading data
                }
            }

        })

        img_profile_edit.setOnClickListener {
            val intent = Intent(activity, EditProfileActivity::class.java)
            startActivity(intent)
        }
    }

    private fun initTablayout() {
        profile_tab_layout.tabGravity = TabLayout.GRAVITY_FILL

        val adapter = ProfilebarAdapter(childFragmentManager)
        adapter.addFragments(UserFavoriteFragment())
        adapter.addFragments(AccountsSettingFragment())

        profile_view_pager.adapter = adapter
        profile_view_pager.offscreenPageLimit = 2

        profile_tab_layout.setupWithViewPager(profile_view_pager)

    }

    private fun checkShippingData() {

        if (!tokenManager.isLoggedIn) return
        if (tokenManager.shippingData2 != null) return
        if (tokenManager.shippingData != null) return

        val intent = Intent(
            activity,
            ShippingActivity::class.java
        )
        startActivity(intent)
    }

}
